crates/relayer-framework-test/src/relayer_mock/contexts/chain.rs

```diff
@@ -0,0 +1,338 @@
+        let client_consensus = self.query_consensus_state_at_height(receiver.clone(), height)?;
+        if client_consensus.check_received(&packet.port_id, &packet.channel_id, &packet.sequence) {
+            return Err(Error::generic(eyre!("chain `{}` got a TimeoutPacket, but client `{}` state received the packet as received", self.name(), receiver)));
+        }
+
+        // Retrieve the current_state and update it with the newly received acknowledgement
+        let mut new_state = self.get_current_state();
+        new_state.update_timeout(packet.port_id, packet.channel_id, packet.sequence);
+
+        // Update the current_state of the Chain
+        self.update_current_state(new_state)?;
+
+        Ok(())
+    }
+
+    pub fn build_send_packet(
+        &self,
+        client_id: ClientId,
+        channel_id: ChannelId,
+        port_id: PortId,
+        timeout_height: MockHeight,
+        timeout_timestamp: MockTimestamp,
+        sequence: Sequence,
+    ) -> PacketKey {
+        PacketKey::new(
+            client_id,
+            channel_id,
+            port_id,
+            sequence,
+            timeout_height,
+            timeout_timestamp,
+        )
+    }
+
+    /// Update the chain's current_state and past_chain_states at the same time to insure
+    /// they are always synchronized.
+    /// The MockChain must have a one and only one ChainState at every height.
+    fn update_current_state(&self, state: State) -> Result<(), Error> {
+        let latest_height = self.get_current_height();
+        let new_height = latest_height.increment();
+
+        // Update current state
+        let mut locked_current_state = self.current_state.acquire_mutex();
+        *locked_current_state = state;
+
+        // Update current height.
+        let mut locked_current_height = self.current_height.acquire_mutex();
+        *locked_current_height = new_height.clone();
+
+        // Timestamp is increased by 1 second when the Height of a chain increases by 1.
+        self.runtime().runtime.clock.increment_millis(1000)?;
+
+        // After inserting the new state in the current_state, update the past_chain_states
+        // at the given height.
+        let mut locked_past_chain_states = self.past_chain_states.acquire_mutex();
+        locked_past_chain_states.insert(new_height, locked_current_state.clone());
+
+        Ok(())
+    }
+
+    /// If the message is a `SendPacket`, update the received packets,
+    /// and add a `RecvPacket` event to the returned array of events.
+    /// If the message is an `AckPacket`, update the received acknowledgment
+    /// packets.
+    /// If the message is an `UpdateClient` update the consensus state.
+    /// When a RecvPacket and AckPacket are received, verify that the client
+    /// state has respectively sent the message and received the message.
+    pub fn process_messages(&self, messages: Vec<MockMessage>) -> Result<Vec<Vec<Event>>, Error> {
```

Comment:
Actually there is one last fix needed for the semantics of `process_messages`: the messages need to be either all being processed successfully at the same height, or all failed.

So this means that methods like `self.receive_packet` should not directly update the chain state. Instead they should take the current pending chain state, and return a new chain state to be committed. The next message handler would take the chain state returned from the previous message handler. And if any message handling failed, the whole pending chain state is discarded.

Reply:
In any case, I will merge this first so that we don't get too many merge conflicts from too many parallel branches.